jumanji/training/networks/graph_coloring/actor_critic.py

```diff
@@ -0,0 +1,293 @@
+    )
+
+
+class GraphColoringTorso(hk.Module):
+    def __init__(
+        self,
+        num_transformer_layers: int,
+        transformer_num_heads: int,
+        transformer_key_size: int,
+        transformer_mlp_units: Sequence[int],
+        name: Optional[str] = None,
+    ):
+        super().__init__(name=name)
+        self.num_transformer_layers = num_transformer_layers
+        self.transformer_num_heads = transformer_num_heads
+        self.transformer_key_size = transformer_key_size
+        self.transformer_mlp_units = transformer_mlp_units
+        self.model_size = transformer_num_heads * transformer_key_size
+
+    def _make_self_attention_mask(self, adj_matrix: chex.Array) -> chex.Array:
+        # Expand on the head dimension.
+        mask = jnp.expand_dims(adj_matrix, axis=-3)
+        return mask
+
+    def embed_nodes(self, nodes_colored: chex.Array) -> chex.Array:
+        node_embeddings = hk.Linear(self.model_size)(
+            nodes_colored[..., None].astype(float)
+        )
+        return node_embeddings
+
+    def embed_colors(self, colors_used: chex.Array) -> chex.Array:
+        embeddings = hk.Linear(self.model_size)(colors_used[..., None].astype(float))
+        return embeddings
+
+    def __call__(self, observation: Observation) -> chex.Array:
+        """
+        Transforms the observation using a series of transformations.
+
+        The observation is composed of the following components:
+        - observation.action_mask: Represents the colors that are available for the current node.
+        - observation.colors: Represents the colors assigned to each node.
+            Nodes without an assigned color are marked with -1.
+        - observation.current_node_index: Represents the node currently being considered.
+
+        The function first determines which colors are used and which nodes are colored.
+        Then it embeds the colors and nodes, and creates a mask for the adjacency matrix.
+        It further creates two masks to track the relation between colors and nodes.
+        Then the function applies a series of transformer blocks to compute:
+            self-attention on nodes and colors and the cross-attention between nodes and colors.
+        Finally, it extracts the embedding for the current node and computes a new embedding.
+
+        Args:
+            observation: The observation to be transformed.
+
+        Returns:
+            new_embedding: The transformed observation.
+        """
+
+        batch_size, num_nodes = observation.colors.shape
+        colors_used = jnp.isin(observation.colors, jnp.arange(num_nodes))
+        color_embeddings = hk.Linear(self.model_size)(
+            colors_used[..., None].astype(float)
+        )  # Shape (batch_size, num_colors, 128)
+
+        nodes_colored = observation.colors >= 0
+
+        color_embeddings = self.embed_colors(
+            colors_used
+        )  # Shape (batch_size, num_colors, 128)
```

Comment:
This does not seem needed since already defined above. I believe we could remove these `embed_colors` and `embed_nodes` methods that only implements one operation anyway.